p2p/src/authenticated/discovery/actors/listener.rs

```diff
@@ -143,19 +166,223 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Network + Rng + CryptoRng + Metri
                     continue;
                 }
             };
-            debug!(ip = ?address.ip(), port = ?address.port(), "accepted incoming connection");
+            debug!(?address, "accepted incoming connection");
+
+            // Cleanup the rate limiters periodically
+            if accepted > CLEANUP_INTERVAL {
+                self.ip_rate_limiter.shrink_to_fit();
+                self.subnet_rate_limiter.shrink_to_fit();
+                accepted = 0;
+            }
+            accepted += 1;
+
+            // Check whether the IP (and subnet) exceeds its rate limit
+            let ip = address.ip();
+            let ip_limited = if self.ip_rate_limiter.check_key(&ip).is_err() {
+                self.handshakes_ip_rate_limited.inc();
+                debug!(?address, "ip exceeded handshake rate limit");
+                true
+            } else {
+                false
+            };
+            let subnet = ip.subnet();
+            let subnet_limited = if self.subnet_rate_limiter.check_key(&subnet).is_err() {
+                self.handshakes_subnet_rate_limited.inc();
+                debug!(?address, "subnet exceeded handshake rate limit");
+                true
+            } else {
+                false
+            };
+
+            // We wait to check whether the handshake is permitted until after updating both the ip
+            // and subnet rate limiters
+            if ip_limited || subnet_limited {
+                continue;
+            }
+
+            // Attempt to reserve a slot for the handshake
```

Comment:
```suggestion
            // Check whether there are too many ongoing handshakes or there is an ongoing handshake
            // to this IP
```

> We now ensure there is at most 1 concurrent handshake per IP!

May as well memorialize that